src/core/border_router/routing_manager.cpp

```diff
@@ -3689,57 +3706,56 @@ void RoutingManager::PdPrefixManager::ProcessPlatformGeneratedRa(const uint8_t *
-    favoredEntry.Clear();
+    VerifyOrExit(mEnabled, LogWarn("Ignore DHCPv6 delegated prefix since PD is disabled."));
 
-    for (const Option &option : aMessage)
-    {
-        DiscoveredPrefixTable::Entry entry;
+    error = Process(nullptr, &aPrefixTableEntry);
 
-        if (option.GetType() != Option::kTypePrefixInfo || !static_cast<const PrefixInfoOption &>(option).IsValid())
-        {
-            continue;
-        }
+exit:
 
-        mNumPlatformPioProcessed++;
-        entry.SetFrom(static_cast<const PrefixInfoOption &>(option));
+    if (error != kErrorNone)
+    {
+        LogCrit("Failed to process DHCPv6 delegated prefix: %s", ErrorToString(error));
+    }
+}
 
-        if (!IsValidPdPrefix(entry.GetPrefix()))
-        {
-            LogWarn("PdPrefixManager: Ignore invalid PIO entry %s", entry.GetPrefix().ToString().AsCString());
-            continue;
-        }
+Error RoutingManager::PdPrefixManager::Process(const RouterAdvert::RxMessage *aMessage,
+                                               const PrefixTableEntry        *aPrefixTableEntry)
+{
+    bool                         currentPrefixUpdated = false;
+    Error                        error                = kErrorNone;
+    DiscoveredPrefixTable::Entry favoredEntry;
 
-        entry.mPrefix.Tidy();
-        entry.mPrefix.SetLength(kOmrPrefixLength);
+    favoredEntry.Clear();
 
-        // The platform may send another RA message to announce that the current prefix we are using is no longer
-        // preferred or no longer valid.
-        if (entry.GetPrefix() == GetPrefix())
-        {
-            currentPrefixUpdated = true;
-            mPrefix              = entry;
-        }
+    // aMessage or aPrefixTableEntry must be different from null
+    if (aMessage != nullptr)
+    {
+        VerifyOrExit(aMessage->IsValid(), error = kErrorParse);
 
-        if (entry.IsDeprecated())
+        for (const Option &option : *aMessage)
         {
-            continue;
-        }
+            DiscoveredPrefixTable::Entry entry;
 
-        // Some platforms may delegate us more than one prefixes. We will pick the smallest one. This is a simple rule
-        // to pick the GUA prefix from the RA messages since GUA prefixes (2000::/3) are always smaller than ULA
-        // prefixes (fc00::/7).
-        if (favoredEntry.GetPrefix().GetLength() == 0 || entry.GetPrefix() < favoredEntry.GetPrefix())
-        {
-            favoredEntry = entry;
+            if (option.GetType() != Option::kTypePrefixInfo || !static_cast<const PrefixInfoOption &>(option).IsValid())
+            {
+                continue;
+            }
+
+            mNumPlatformPioProcessed++;
+            entry.SetFrom(static_cast<const PrefixInfoOption &>(option));
+            currentPrefixUpdated |= ProcessPrefixEntry(entry, favoredEntry);
         }
     }
+    else // aPrefixTableEntry != nullptr
+    {
+        DiscoveredPrefixTable::Entry entry;
+        entry.SetFrom(*aPrefixTableEntry);
```

Comment:
Newline after variable definition within a `{}` scope?

```suggestion
        DiscoveredPrefixTable::Entry entry;
        
        entry.SetFrom(*aPrefixTableEntry);
```